keps/sig-network/1111-dynamically-re-configure-lb-node-set/README.md

```diff
@@ -0,0 +1,681 @@
+why should the state of the node / Kubelet impact any routing decision? Either
+the proxy is running fine and knows the endpoint is running on the node, or not.
+The only reason to include such logic would be to try to perform pre-mature
+optimization, by trying to predict that if the Kubelet / node is not running
+fine right now, then maybe the endpoint will be scheduled elsewhere soon. But
+that can be very wrong and cut ingress traffic to the application when the node
+experiences short transient errors which will not cause the endpoint to be
+re-scheduled at all.
+
+Given this, it is therefor the opinion of this KEP that the health of the node
+included in the application load balancer's backend pool, should be: the health
+of service proxy.
+
+### Extend `healthCheckNodePort` to class A services
+
+The following addresses [aligning health checks between services](#align-health-checks-between-services)
+
+Given that Kubernetes already defines a port for the load balancer HC to target
+for class B type services, and the fact that we don't want to re-configure the
+HC when the service changes to class A: it would make sense to extend the
+`healthCheckNodePort` field to class A services. Once the service object is
+updated, the existing value would continue to be used.
+
+The name of this field is also rather representative of what it does, which
+would help user comprehension for what is to be expected of this field. With the
+current HC probes defined individually by each cloud provider it takes the user
+some serious spelunking to figure out exactly why a node has been added/removed
+from the LB set. `healthCheckNodePort` has a very narrow scope (compared to
+other HC endpoints in use today, such as: `NodePort` or the Kubelet's healthz
+port). Besides allowing us to document it better and provide the user with a
+better user experience in general, we would also avoiding mixing different
+features depending on specific fields (ex: `NodePort` - which serves as a
+back-bone to a lot of features) and use it as a "dedicated HC intended
+attribute". Given that we would extend it, we also would not need to worry about
+backport/version skew compatibility - which would have be the case should we opt
+for another mechanism which might impact existing class B functionality
+dependent upon `healthCheckNodePort`.
+
+Clearly a big difficulty with doing this will be to make sure that whatever is
+configured for class B services, remains to only be configured for class B
+services - even though both classes of services will have this field.
+
+Doing this would require: changing the API for class A services, change the HCs
+configured by cloud providers to point to the `healthCheckNodePort` for all
+services, and change the service proxies to open ports for all services which
+have this field, but respond differently depending on if the service is class A
+or B. If A: check healthz, if B: check healthz and if there's a local endpoint.
```

Comment:
It has to be a criteria of graduation that cloud providers are using this new `healthCheckNodePort`, modifying the API if is not being used is a tremendous effort

Reply:
I know the major public cloud providers all use it. Not sure about _all_ though.

Reply:
yeah, I think that the 3 majors should be a good indicator